src/main/java/org/springframework/data/mapping/model/KotlinValueUtils.java

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.mapping.model;
+
+import kotlin.jvm.JvmClassMappingKt;
+import kotlin.jvm.internal.Reflection;
+import kotlin.reflect.KCallable;
+import kotlin.reflect.KClass;
+import kotlin.reflect.KFunction;
+import kotlin.reflect.KParameter;
+import kotlin.reflect.KProperty;
+import kotlin.reflect.KType;
+import kotlin.reflect.KTypeParameter;
+import kotlin.reflect.jvm.ReflectJvmMapping;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.springframework.core.KotlinDetector;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * Utilities for Kotlin Value class support.
+ *
+ * @author Mark Paluch
+ * @since 3.2
+ */
+class KotlinValueUtils {
+
+	/**
+	 * Returns whether the given {@link KType} is a {@link KClass#isValue() value} class.
+	 *
+	 * @param type the kotlin type to inspect.
+	 * @return {@code true} the type is a value class.
+	 */
+	public static boolean isValueClass(KType type) {
+		return type.getClassifier()instanceof KClass<?> kc && kc.isValue();
+	}
+
+	/**
+	 * Returns whether the given class makes uses Kotlin {@link KClass#isValue() value} classes.
+	 *
+	 * @param type the kotlin type to inspect.
+	 * @return {@code true} when at least one property uses Kotlin value classes.
+	 */
+	public static boolean hasValueClassProperty(Class<?> type) {
+
+		if (!KotlinDetector.isKotlinType(type)) {
+			return false;
+		}
+
+		KClass<?> kotlinClass = JvmClassMappingKt.getKotlinClass(type);
+
+		for (KCallable<?> member : kotlinClass.getMembers()) {
+			if (member instanceof KProperty<?> kp && isValueClass(kp.getReturnType())) {
```

Comment:
@mp911de Sorry for trespassing, but I believe this might work incorrectly in case there are unrelated properties in the class. For example:
```
@JvmInline
value class Z(val value: String)

data class D(val x: Int) {
    val unrelated1 = Z("1")

    val Any.unrelated2: Z get() = Z("2")
}
```

`hasValueClassProperty` will return true for the class `D`, but it probably should be false (if I understood call sites correctly).

I suppose you're interested in finding only if any property in the primary constructor has a value class type? For that you need to take only non-extension properties (i.e. [`KClass.memberProperties`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect.full/member-properties.html)) whose names are equal to the parameters of the primary constructor.